### PR TITLE
feat(config): session correlation header injection configuration

### PR DESCRIPTION
**Pull request by SasSwart**
## PR Map

1. #196
2. 👉🏼 **#197**
3. #198

---

RFC: [Bridge ↔ Boundaries Correlation](https://www.notion.so/Bridge-Boundaries-Correlation-313d579be59281f3b4efdbfd6896775a)

Add YAML and CLI configuration surface for session correlation header injection per FR 2 of the RFC.

Depends on #196.

## Changes

- **`config/session_correlation.go`**: New file. Defines `SessionCorrelationConfig`, `InjectTarget`, `ParseInjectTarget` (with duplicate-key detection), `ValidateSessionCorrelation`, fixed header name constants (`SessionIDHeaderName`, `SequenceNumberHeaderName`), and `DefaultInjectTargetFromEnv` — a helper that derives a default inject target from `CODER_AGENT_URL` when no explicit targets are configured.
- **`config/session_correlation_test.go`**: New file. Table-driven tests for `ParseInjectTarget` (including duplicate-key rejection), `ValidateSessionCorrelation`, the end-to-end `NewAppConfigFromCliConfig` session correlation path (with controlled `environ`), `DefaultInjectTargetFromEnv`, CLI+YAML target merging (`TestBuildSessionCorrelation_TargetMerge`), and the `CODER_AGENT_URL` fallback path (`TestBuildSessionCorrelation_AgentURLFallback`).
- **`config/config.go`**: Wire session correlation fields into `CliConfig` (three new fields for serpent bindings: `SessionCorrelationEnabled`, `InjectSessionIDTarget`, `InjectSessionIDTargets`) and `AppConfig` (new `SessionCorrelation` field). `NewAppConfigFromCliConfig` now accepts an `environ []string` parameter so the `CODER_AGENT_URL` fallback is testable without relying on the host environment. Add `buildSessionCorrelation` helper that merges YAML+CLI inject targets, falls back to a target derived from `CODER_AGENT_URL` when none are configured and session correlation is enabled, and validates the result.
- **`cli/cli.go`**: Register three new serpent options, pass `os.Environ()` to `NewAppConfigFromCliConfig`, and add usage examples including a zero-config workspace example.

## New CLI flags

| Flag | Env | YAML | Default | Description |
|---|---|---|---|---|
| `--enable-session-correlation` | `BOUNDARY_SESSION_CORRELATION_ENABLED` | `session_correlation_enabled` | `false` | Toggle; when enabled with no explicit targets, the inject target is auto-derived from `CODER_AGENT_URL` (set automatically in Coder workspaces) |
| `--session-id-inject-target` | `BOUNDARY_SESSION_ID_INJECT_TARGET` | — | — | Repeatable inject target (one target per flag invocation; env accepts one value; use YAML for multiple). Format: `domain=<host> [path=<glob>]` |
| — | — | `session_id_inject_targets` | — | YAML-only list of inject targets |

Header names are fixed constants (`X-Coder-Agent-Firewall-Session-Id`, `X-Coder-Agent-Firewall-Sequence-Number`) matching what Coder AI Gateway expects — they are not user-configurable.

Config validation ensures that when correlation is enabled at least one inject target is present (either explicit or auto-derived from `CODER_AGENT_URL`). Parsing validates the `domain=... path=...` key-value format, rejects unknown keys, and rejects duplicate keys within a single target string.

This commit adds config and validation only; runtime injection is wired in a follow-up PR.

> [!NOTE]
> This PR was authored by Coder Agents.

### CODE REVIEW

File: cli/cli.go [hunk 1]
```diff
@@ -37,7 +37,22 @@ func NewCommand(version string) *serpent.Command {
...
+  # Block everything by default (implicit)
+
+  # Enable session correlation inside a Coder workspace (inject target auto-derived from CODER_AGENT_URL)
+  boundary --enable-session-correlation \
+    --allow "domain=dev.coder.com" -- python train.py
+
+  # Enable session correlation with an explicit inject target (e.g. outside a workspace or custom deployment)
+  boundary --enable-session-correlation \
+    --session-id-inject-target "domain=mydeployment.coder.com path=/api/v2/aibridge/*" \
+    --allow "domain=mydeployment.coder.com" -- python train.py
```
Comment (ssncferreira) [inline cli/cli.go:49]: Would it make sense to set automatically `--session-id-inject-target` with whatever domain is set via `--allow`?

File: cli/cli.go [hunk 2]
```diff
@@ -169,6 +184,43 @@ func BaseCommand(version string) *serpent.Command {
...
+				Env:         "BOUNDARY_SESSION_ID_INJECT_TARGET",
+				Description: `Inject target (repeatable via flag; env accepts one value). Requests matching these targets receive session correlation headers. For multiple targets use the YAML config. Format: "domain=<host> [path=<glob>]".`,
+				Value:       &cliConfig.InjectSessionIDTarget,
+				YAML:        "", // CLI only, YAML uses session_id_inject_targets.
+			},
+			{
+				Flag:        "", // No CLI flag, YAML only.
+				Description: "Inject targets from config file (YAML only).",
+				Value:       &cliConfig.InjectSessionIDTargets,
+				YAML:        "session_id_inject_targets",
```
Comment (ssncferreira) [inline cli/cli.go:206]: Couldn't this be merged with the CLI `session-id-inject-target`? Is this a common pattern in coder to have separate CLI vs YAML options?

File: config/config.go [hunk 3]
```diff
@@ -108,6 +120,12 @@ func NewAppConfigFromCliConfig(cfg CliConfig, targetCMD []string) (AppConfig, er
 
 	userInfo := GetUserInfo()
 
+	// Build session correlation config from CLI and YAML sources.
+	sc, err := buildSessionCorrelation(cfg, os.Environ())
```
Comment (ssncferreira) [inline config/config.go]: nit: `os.Environ()` is hardcoded here, making `NewAppConfigFromCliConfig` untestable for the `CODER_AGENT_URL` fallback path 🤔  Maybe we can pass `environ []string` as a parameter and use `os.Environ()` from the caller?

File: config/config.go [hunk 4]
```diff
@@ -108,6 +120,12 @@ func NewAppConfigFromCliConfig(cfg CliConfig, targetCMD []string) (AppConfig, er
...
+	if err != nil {
+		return AppConfig{}, fmt.Errorf("session correlation config: %w", err)
```
Comment (ssncferreira) [inline config/config.go:123]: nit: This should hint at the fix, e.g. `"...set --session-id-inject-target or ensure CODER_AGENT_URL is set"`.

File: config/session_correlation.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	var target InjectTarget
+	for _, part := range strings.Fields(raw) {
+		key, value, ok := strings.Cut(part, "=")
+		if !ok {
+			return InjectTarget{}, fmt.Errorf(
+				"inject target: malformed key-value pair %q, expected key=value", part,
+			)
+		}
+		switch key {
+		case "domain":
```
Comment (ssncferreira) [inline config/session_correlation.go:77]: There is an edge case here, imagine `--session-id-inject-target "domain=staging.coder.com domain=prod.coder.com"`, domain will be overwritten. Probably worth checking for duplicates and adding a test case for this

File: config/session_correlation.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+		return nil
+	}
+
+	u, err := url.Parse(raw)
+	if err != nil || u.Host == "" {
+		return nil
+	}
+
+	return &InjectTarget{
+		Domain: u.Hostname(),
```
Comment (ssncferreira) [inline config/session_correlation.go:123]: IIUC inject target basically defines which outbound HTTP requests should receive the session correlation headers, but we are stripping the port here. What happens if coder runs on a non-standard port?

File: config/config.go [hunk 7]
```diff
@@ -122,5 +140,54 @@ func NewAppConfigFromCliConfig(cfg CliConfig, targetCMD []string) (AppConfig, er
...
+	}
+
+	if len(targets) == 0 && cfg.SessionCorrelationEnabled.Value() {
+		if t := DefaultInjectTargetFromEnv(environ); t != nil {
+			targets = []InjectTarget{*t}
+		}
+	}
+
+	// Apply defaults for header names.
+	sessionIDHeader := cfg.SessionIDHeaderName.Value()
```
Comment (ssncferreira) [inline config/config.go]: Do we want to make the headers configurable? This means a user could set existing headers for the correlation, like the "Authorization" header, for instance. Also, since this is for correlation with Coder, I assume we would also need to set this on Coder side. So it makes sense to have an agreement on both parts for the same header value, right?

File: config/session_correlation_test.go [hunk 8]
```diff
@@ -0,0 +1,465 @@
...
+				t.Errorf("Domain: got %q, want %q", got.Domain, tc.want.Domain)
+			}
+			if got.Path != tc.want.Path {
+				t.Errorf("Path: got %q, want %q", got.Path, tc.want.Path)
+			}
+		})
+	}
+}
+
+func TestBuildSessionCorrelation_AgentURLFallback(t *testing.T) {
```
Comment (ssncferreira) [inline config/session_correlation_test.go:427]: I think we are missing a test for merging CLI and YAML configs.

File: cli/cli.go (hunk 1 above)
Comment (SasSwart) [reply cli/cli.go:49]: Unfortunately not. `--allow` might point to github.com, for example, which should not receive this header. We do however automatically infer a value for the target based on [this discussion](https://www.notion.so/Boundary-Bridge-Correlation-31ad579be592803aa8b3d48348ccdde9?source=copy_link#352d579be59280a09eb3f5cb8ea412b0)

File: cli/cli.go (hunk 2 above)
Comment (SasSwart) [reply cli/cli.go:206]: This is the established pattern in the current codebase. See `--allow` above. We have to do this because repeatable flags and yaml unmarshalling conflict when applied to the same flag.

File: config/session_correlation.go (hunk 6 above)
Comment (SasSwart) [reply config/session_correlation.go:123]: If coderd runs on a non-standard port, it will still get the header. The rules engine already strips the port in its evaluations. A side-effect is that *all* ports on this domain will get the header. This is slightly broader than defined in the RFC, but I think its a practical tradeoff. To be port specific would require a change to the rules engine, which I do not want to include in the scope of this PR. Happy to do it as a follow-up.

File: config/config.go (hunk 7 above)
Comment (SasSwart) [reply config/config.go]: You're right. Firewall and Gateway should definitely agree on the header name for correlation to work. Both should use the same default header name. 

I do think we should keep it configurable on both ends as a safe bet against unknowns in customer environments that may require specific headers. 

Your point about setting the name to conflict with existing headers is a good one. Firewall should never overwrite headers, so if an operator configures it to use a well-known header like `Authorization`, Firewall should refuse to write it and log the error.

File: config/config.go (hunk 3 above)
Comment (SasSwart) [reply config/config.go]: done!

File: config/session_correlation.go (hunk 5 above)
Comment (SasSwart) [reply config/session_correlation.go:77]: --session-id-inject-target "domain=staging.coder.com domain=prod.coder.com" is not valid. Rather, it should be --session-id-inject-target "domain=staging.coder.com"  --session-id-inject-target "domain=prod.coder.com". I'll add a test case, followed by validation and documentation.

Comment (SasSwart) [reply config/session_correlation.go:77]: done

File: config/session_correlation_test.go (hunk 8 above)
Comment (SasSwart) [reply config/session_correlation_test.go:427]: added

File: config/session_correlation.go (hunk 5 above)
Comment (SasSwart) [reply config/session_correlation.go:77]: done!

**Review by SasSwart — COMMENTED** [review]
Most review feedback tended to. One tricky outstanding comment left.

File: config/config.go (hunk 4 above)
Comment (SasSwart) [reply config/config.go:123]: done!

File: config/config.go (hunk 7 above)
Comment (SasSwart) [reply config/config.go]: I've removed the ability to configure these names. Thanks for the consult!

File: config/session_correlation_test.go [hunk 9]
```diff
@@ -0,0 +1,521 @@
...
+			wantErr: true,
+		},
+		{
+			name:    "unknown key",
+			input:   "domain=example.com port=443",
+			wantErr: true,
+		},
+		{
+			name:    "duplicate domain key",
+			input:   "domain=staging.coder.com domain=prod.coder.com",
```
Comment (ssncferreira) [inline config/session_correlation_test.go:63]: Nice 🤩

**Review by ssncferreira — APPROVED** [review]
LGTM 👍 Thank you for addressing the comments!

File: cli/cli.go (hunk 2 above)
Comment (evgeniy-scherbina) [reply cli/cli.go:206]: IIRC `--allow` is an exception, not a rule. It was done to make sure that you can combine yaml config + CLI flags.

Yaml config contains primary configuration, `--allow` flag allows to append few extra domains without updating config file. So it was done for convenience.

If it's possible one flag is preferable.

File: config/session_correlation.go [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+	// CoderAgentURLEnv is the environment variable set by the Coder workspace
+	// agent that points to the control plane. Boundary uses it to derive a
+	// default inject target when none is explicitly configured.
+	CoderAgentURLEnv = "CODER_AGENT_URL"
+)
+
+// InjectTarget represents a parsed target for session correlation header
+// injection. Requests matching the domain (and optional path glob) will
+// receive the session ID and sequence number headers.
+type InjectTarget struct {
```
Comment (evgeniy-scherbina) [inline config/session_correlation.go:33]: I'm a bit concerned that `--session-id-inject-target(s)` syntax similar but not identical to `--allow` syntax; and that parsing & matching logic isn't reused (see rulesengine pkg).

rulesengine pkg is well documented, so customers will probably rely on it; and it will be confusing that `--session-id-inject-target(s)` syntax different from it.
